File: python/tests/ral_object_test.py


from openral_py.ral_object import RalObject


class TestRalObject:
    
    def test_from_map(self):
        #Arrange
        data = {
            "identity": {
                "UID": "uid",
                "name": "thing",
                "siteTag": "siteTag",
                "alternateIDs": [],
                "alternateNames": []
            },
            "currentOwners": [],
            "definition": {
                "definitionText": "An object or entity that is not or cannot be named specifically",
                "definitionURL": "https://www.thefreedictionary.com/thing"
            },
            "objectState": "undefined",
            "template": {
                "RALType": "thing",
                "version": "1",
                "objectStateTemplates": "generalObjectState"
            },
            "specificProperties": [
                {
                "key": "serial number",
                "value": "my_value",
                "unit": "String"
                }
            ],
            "currentGeolocation": {
                "container": {
                "UID": "unknown"
                },
                # todo
                # "postalAddress": {
                # "country": "unknown",
                # "cityName": "unknown",
                # "cityNumber": "unknown",
                # "streetName": "unknown",
                # "streetNumber": "unknown"
                # },
                # "3WordCode": "unknown",
                # "geoCoordinates": {
                # "longitude": 0,
                # "latitude": 0
                # },
                # "plusCode": "unknown"
            },
            "locationHistoryRef": [],
            "ownerHistoryRef": [],
            "methodHistoryRef": [],
            "linkedObjectRef": []
        }

        #Act
        ral_object = RalObject.from_map(data)

        #Assert
        assert ral_object.identity.uid == "uid"
        assert ral_object.identity.name == "thing"
        assert ral_object.identity.site_tag == "siteTag"
        assert ral_object.identity.alternate_ids == []
        assert ral_object.identity.alternate_names == []

        assert ral_object.current_owners == []

        assert ral_object.definition.definition_text == "An object or entity that is not or cannot be named specifically"
        assert ral_object.definition.definition_url == "https://www.thefreedictionary.com/thing"

        assert ral_object.object_state == "undefined"

        assert ral_object.template.ral_type == "thing"
        assert ral_object.template.version == "1"
        assert ral_object.template.object_state_templates == "generalObjectState"


        specific_properties = ral_object.specific_properties.get("serial number")
        if specific_properties is not None:
            assert specific_properties.key == "serial number"
            assert specific_properties.value == "my_value"
            assert specific_properties.unit == "String"
        else:
            assert False, "specific_properties is None"

        container = ral_object.current_geo_location.container
        if container is not None:
            assert container.uid == "unknown"
        else:
            assert False, "container is None"

        # todo
        # assert ral_object.current_geolocation.postal_address.country == "unknown"
        # assert ral_object.current_geolocation.postal_address.city_name == "unknown"
        # assert ral_object.current_geolocation.postal_address.city_number == "unknown"
        # assert ral_object.current_geolocation.postal_address.street_name == "unknown"
        # assert ral_object.current_geolocation.postal_address.street_number == "unknown"

        assert ral_object.location_history_ref == []
        assert ral_object.owner_history_ref == []
        assert ral_object.method_ristory_ref == []
        assert ral_object.linked_object_ref == []
    
    
    def test_from_map_minimum_data(self):
        #Arrange
        data = {
            "identity": {
                "UID": "uid",
            },
            "definition": {
                
            },
            
            "template": {
                "RALType": "thing",
                "version": "1",
                "objectStateTemplates": "generalObjectState"
            },
            "specificProperties": [],
            "currentGeolocation": {
                
                # todo
                # "postalAddress": {
                # "country": "unknown",
                # "cityName": "unknown",
                # "cityNumber": "unknown",
                # "streetName": "unknown",
                # "streetNumber": "unknown"
                # },
                # "3WordCode": "unknown",
                # "geoCoordinates": {
                # "longitude": 0,
                # "latitude": 0
                # },
                # "plusCode": "unknown"
            },
        }

        #Act
        ral_object = RalObject.from_map(data)

        #Assert
        assert ral_object.identity.uid == "uid"
        assert ral_object.identity.name is None
        assert ral_object.identity.site_tag is None
        assert ral_object.identity.alternate_ids == []
        assert ral_object.identity.alternate_names == []

        assert ral_object.current_owners == []

        assert ral_object.definition.definition_text is None
        assert ral_object.definition.definition_url is None

        assert ral_object.object_state == "undefined", "object_state default value is undefined"

        assert ral_object.template.ral_type == "thing"
        assert ral_object.template.version == "1"
        assert ral_object.template.object_state_templates == "generalObjectState"


        assert len(ral_object.specific_properties.map) == 0, "specific_properties is empty"

        container = ral_object.current_geo_location.container
        assert container is None, "container is None"

        # todo
        # assert ral_object.current_geolocation.postal_address.country == "unknown"
        # assert ral_object.current_geolocation.postal_address.city_name == "unknown"
        # assert ral_object.current_geolocation.postal_address.city_number == "unknown"
        # assert ral_object.current_geolocation.postal_address.street_name == "unknown"
        # assert ral_object.current_geolocation.postal_address.street_number == "unknown"

        assert ral_object.location_history_ref == [], "locationHistoryRef is empty list"
        assert ral_object.owner_history_ref == [], "ownerHistoryRef is empty list"
        assert ral_object.method_ristory_ref == [], "methodHistoryRef is empty list"
        assert ral_object.linked_object_ref == [], "linkedObjectRef is empty list"
    

File: dart/test/model/ral_object_test.dart
import 'package:openral_core/src/model/container.dart';
import 'package:openral_core/src/model/current_geo_location.dart';
import 'package:openral_core/src/model/definition.dart';
import 'package:openral_core/src/model/identity.dart';
import 'package:openral_core/src/model/object_ref.dart';
import 'package:openral_core/src/model/object_template.dart';
import 'package:openral_core/src/model/ral_object.dart';
import 'package:openral_core/src/model/specific_properties.dart';
import 'package:openral_core/src/model/specific_property.dart';
import 'package:test/test.dart';

void main() {
  group(
    RalObject,
    () {
      group(
        'fromMap',
        () {
          test('should return a RalObject with the correct values', () {
            final data = {
              "identity": {
                "UID": "1234",
                "name": "thing",
                "siteTag": "",
                "alternateIDs": ["9876"],
                "alternateNames": ["myName"],
              },
              "currentOwners": [
                {"UID": "ownerUid", "role": "ownerRole"}
              ],
              "definition": {
                "definitionText": "An object or entity that is not or cannot be named specifically",
                "definitionURL": "https://www.thefreedictionary.com/thing"
              },
              "objectState": "undefined",
              "template": {
                "RALType": "thing",
                "version": "1",
                "objectStateTemplates": ["generalObjectState"],
              },
              "specificProperties": [
                {
                  "key": "serial number",
                  "value": "",
                  "unit": "String",
                }
              ],
              "currentGeolocation": {
                "container": {
                  "UID": "unknown",
                },
                "postalAddress": {
                  "country": "unknown",
                  "cityName": "unknown",
                  "cityNumber": "unknown",
                  "streetName": "unknown",
                  "streetNumber": "unknown",
                },
                "3WordCode": "unknown",
                "geoCoordinates": {
                  "longitude": 0,
                  "latitude": 0,
                },
                "plusCode": "unknown"
              },
              "locationHistoryRef": [],
              "ownerHistoryRef": [],
              "methodHistoryRef": [],
              "linkedObjectRef": [
                {"UID": "linkedObjectUid", "role": "linkedObjectRole"}
              ]
            };

            final result = RalObject.fromMap(data);

            expect(result.isLeft, isTrue, reason: result.isRight ? result.right.toString() : null);

            final ralObject = result.left;

            expect(ralObject.identity.uid, equals("1234"));
            expect(ralObject.identity.name, equals("thing"));
            expect(ralObject.identity.siteTag, equals(""));
            expect(ralObject.identity.alternateIDs, equals(["9876"]));
            expect(ralObject.identity.alternateNames, equals(["myName"]));

            expect(ralObject.currentOwners.length, equals(1), reason: "currentOwners should contain one element");
            expect(ralObject.currentOwners[0].uid, equals("ownerUid"));

            expect(ralObject.definition.definitionText, equals("An object or entity that is not or cannot be named specifically"));
            expect(ralObject.definition.definitionURL, equals("https://www.thefreedictionary.com/thing"));

            expect(ralObject.objectState, equals("undefined"));

            expect(ralObject.template.ralType, equals("thing"));
            expect(ralObject.template.version, equals("1"));
            expect(ralObject.template.objectStateTemplates, equals(["generalObjectState"]));

            expect(
              ralObject.specificProperties.containsKey("serial number"),
              isTrue,
              reason: "specificProperties should contain the key 'serial number'",
            );

            expect(ralObject.currentGeoLocation.container!.uid, equals("unknown"));
            //TODO: Other geo location fields

            expect(ralObject.locationHistoryRef, equals([]), reason: "locationHistoryRef should be empty");
            expect(ralObject.ownerHistoryRef, equals([]), reason: "ownerHistoryRef should be empty");
            expect(ralObject.methodHistoryRef, equals([]), reason: "methodHistoryRef should be empty");

            expect(ralObject.linkedObjectRef.length, equals(1), reason: "linkedObjectRef should contain one element");
            expect(ralObject.linkedObjectRef[0].uid, equals("linkedObjectUid"));
          });
        },
      );

      group('toMap', () {
        test('returns map with values', () async {
          //arrange
          final ralObject = RalObject(
            identity: Identity(
              uid: "uid1",
              name: "name1",
              siteTag: "siteTag1",
              alternateIDs: ["alternateID1"],
              alternateNames: ["alternateName1"],
            ),
            currentOwners: [
              ObjectRef(
                uid: 'ownerUid',
                role: 'ownerRole',
              ),
            ],
            definition: Definition(
              definitionText: "definitionText1",
              definitionURL: "definitionURL1",
            ),
            objectState: "objectState1",
            template: ObjectTemplate(
              ralType: "ralType1",
              version: "version1",
              objectStateTemplates: ["objectStateTemplates1"],
            ),
            specificProperties: SpecificProperties({
              "key1": SpecificProperty(
                key: "key1",
                value: "value1",
                unit: "unit1",
              ),
              "key2": SpecificProperty(
                key: "key2",
                value: "value2",
                unit: "unit2",
              ),
            }),
            currentGeoLocation: CurrentGeoLocation(
              container: Container(
                uid: "uid1",
              ),
            ),
            locationHistoryRef: ["locationHistoryRef1"],
            ownerHistoryRef: ["ownerHistoryRef1"],
            methodHistoryRef: ["methodHistoryRef1"],
            linkedObjectRef: [
              ObjectRef(
                uid: 'linkedObjectUid',
                role: 'linkedObjectRole',
              ),
            ],
          );

          //act
          final map = ralObject.toMap();

          //assert
          expect(
            map,
            equals({
              "identity": {
                "UID": "uid1",
                "name": "name1",
                "siteTag": "siteTag1",
                "alternateIDs": ["alternateID1"],
                "alternateNames": ["alternateName1"],
              },
              "currentOwners": [
                {"UID": "ownerUid", "role": "ownerRole"}
              ],
              "definition": {
                "definitionText": "definitionText1",
                "definitionURL": "definitionURL1",
              },
              "objectState": "objectState1",
              "template": {
                "RALType": "ralType1",
                "version": "version1",
                "objectStateTemplates": ["objectStateTemplates1"],
              },
              "specificProperties": [
                {
                  "key": "key1",
                  "value": "value1",
                  "unit": "unit1",
                },
                {
                  "key": "key2",
                  "value": "value2",
                  "unit": "unit2",
                },
              ],
              "currentGeolocation": {
                "container": {
                  "UID": "uid1",
                },
                // uncomment when implemented
                // "postalAddress": {
                //   "country": "unknown",
                //   "cityName": "unknown",
                //   "cityNumber": "unknown",
                //   "streetName": "unknown",
                //   "streetNumber": "unknown",
                // },
                // "3WordCode": "unknown",
                // "geoCoordinates": {
                //   "longitude": 0,
                //   "latitude": 0,
                // },
                // "plusCode": "unknown"
              },
              "locationHistoryRef": ["locationHistoryRef1"],
              "ownerHistoryRef": ["ownerHistoryRef1"],
              "methodHistoryRef": ["methodHistoryRef1"],
              "linkedObjectRef": [
                {
                  "UID": "linkedObjectUid",
                  "role": "linkedObjectRole",
                },
              ],
            }),
          );
        });
      });
    },
  );
}
